### ISSUE TITLE
torch.nn.functional.cross_entropy different loss when providing one_hot_target and class weights

### ISSUE THREAD

**Issue by Infrared1029** [pytorch/pytorch#107680]
### 🐛 Describe the bug

```python
inputs = torch.tensor([[2.3, 2.1, 4.5],
                         [4.2, 3.2, 5.1]])

target = torch.tensor([1, 2])
one_hot_target = torch.tensor([[0,1,0], [0,0,1]]).float()

weights = torch.tensor([25., 25., 100.])

torch.nn.functional.cross_entropy(inputs, target, reduction="mean", weight=weights) #output: 0.8705
torch.nn.functional.cross_entropy(inputs, one_hot_target, reduction="mean", weight=weights) #output: 54.4052
```

This seems to only occur with `mean` reduction when `weight` is provided, `sum` and `none` output the same thing.

### Versions

PyTorch version: 2.0.1+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux 10 (buster) (x86_64)
GCC version: (Debian 8.3.0-6) 8.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.28

Python version: 3.10.0 (default, Mar  3 2022, 09:58:08) [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.15.0-1041-azure-x86_64-with-glibc2.28
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
Address sizes:       48 bits physical, 48 bits virtual
CPU(s):              8
On-line CPU(s) list: 0-7
Thread(s) per core:  2
Core(s) per socket:  4
Socket(s):           1
NUMA node(s):        1
Vendor ID:           AuthenticAMD
CPU family:          25
Model:               1
Model name:          AMD EPYC 7763 64-Core Processor
Stepping:            1
CPU MHz:             2445.425
BogoMIPS:            4890.85
Virtualization:      AMD-V
Hypervisor vendor:   Microsoft
Virtualization type: full
L1d cache:           32K
L1i cache:           32K
L2 cache:            512K
L3 cache:            32768K
NUMA node0 CPU(s):   0-7
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl tsc_reliable nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy svm cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw topoext invpcid_single vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves clzero xsaveerptr rdpru arat npt nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold v_vmsave_vmload umip vaes vpclmulqdq rdpid fsrm

Versions of relevant libraries:
[pip3] numpy==1.25.2
[pip3] torch==2.0.1+cpu
[pip3] torch-scatter==2.1.1
[conda] numpy                     1.25.2                   pypi_0    pypi

**Comment by ptrblck** [comment]
I believe the issue was explained [here](https://discuss.pytorch.org/t/cce-loss-mean-reduction-differs-between-sparse-and-dense-labels/185464/2?u=ptrblck) and also discussed [here](https://github.com/pytorch/pytorch/issues/61309) and [here](https://github.com/pytorch/pytorch/pull/61044).
